target/linux/mpc85xx/image/p1010.mk

```diff
@@ -17,10 +17,32 @@ define Device/tplink_tl-wdr4900-v1
   SUPPORTED_DEVICES += tl-wdr4900-v1
   ARTIFACTS := fdt.bin
   ARTIFACT/fdt.bin := append-dtb
-  IMAGES := factory.bin sysupgrade.bin
+  IMAGES := sysupgrade.bin #Factory wont work
   IMAGE/sysupgrade.bin := tplink-v1-image sysupgrade | append-metadata
-  IMAGE/factory.bin := tplink-v1-image factory
-  DEFAULT := n
+  DEVICE_COMPAT_VERSION := 2.0
+  DEVICE_COMPAT_MESSAGE := \n$\
+    !The kernel size of the stock TP-Link TL-WDR4900 v1 is too small! \n$\
+    To upgrade, please take a look at the install instructions over \
+    at the device's wiki: <https://openwrt.org/toh/tp-link/tl-wdr4900> \n$\
+    Run the following script into a shell on the device and retry this \
+    sysupgrade again: \n$\
+    cat <<- "EOF" > /tmp/uboot-fix.sh; sh /tmp/uboot-fix.sh \n$\
+    . /lib/functions.sh \n$\
+    . /lib/functions/system.sh \n$\
+    opkg update && opkg install kmod-mtd-rw || exit 2 \n$\
+    insmod mtd-rw i_want_a_brick=y || exit 3 \n$\
+    cp "/dev/mtd$$$$(find_mtd_index 'u-boot')" /tmp/uboot \n$\
+    cp /tmp/uboot /tmp/uboot_patched \n$\
+    strings -td < /tmp/uboot | grep '^ *[0-9]* spiboot=' | \n$\
+       awk '{print $$$$1}' | \n$\
+       while read offset; do \n$\
+         echo -n "spiboot=sf probe 0;sf read 2000000 60200 69f000;" | dd of=/tmp/uboot_patched bs=1 seek=$$$${offset} conv=notrunc \n$\
```

Comment:
If we have to patch, I would prefer not to hard-code the `bootcmd`, but instead, refer to a u-boot variable, and then `fw_setenv` that variable appropriately. (We would first of course need to make the u-boot env partition variable writable.)

Reply:
This device have only hardcoded envs and no 'saveenv'  cmd. Tp-link offen offer u-boot without env partition. How to find in binary if env partition creation is possible?

Reply:
I modified and compiled uboot, adding env support and several commands (cmp, cp, echo, nm, mm, mw, test, unzip...).
[wdr4900v1_uboot_mod.zip](https://github.com/openwrt/openwrt/files/9690145/wdr4900v1_uboot_mod.zip)

Uboot uses ~221KB, since we have 320KB for it there's no problem using the last 64KB for the environment, tested here, everything works.
Openwrt detection can be easily done now by searching the "Openwrt" string at the start of the kernel:
    setenv openwrt_str_load 'mw.b 2000000 4f;mw.b 2000001 70;mw.b 2000002 65;mw.b 2000003 6e;mw.b 2000004 57;mw.b 2000005 72;mw.b 2000006 74;mw.b 2000007 20;sf
[wdr4900v1_uboot_mod.zip](https://github.com/openwrt/openwrt/files/9690145/wdr4900v1_uboot_mod.zip)
 probe 0;sf read 2000008 60228 8'
    setenv openwrt_detect 'run openwrt_str_load;if cmp.b 2000000 2000008 8;then echo Openwrt found; else echo Openwrt not found;fi'

Then just do "run openwrt_detect" at boot.

If anyone wants to compile himself, just replace P1010RDB.h at
>> TL-WDR4900_eu_1.0_GPL/p1010/uboot/u-boot-2010.12/include/configs
With this one:
[P1010RDB.zip](https://github.com/openwrt/openwrt/files/9690157/P1010RDB.zip)
Remember to append the uboot image to the boot header!
[wdr4900v1_uboot_header.zip](https://github.com/openwrt/openwrt/files/9690170/wdr4900v1_uboot_header.zip)
 
Ex. 
    cat header.bin uboot.bin > wdr4900_uboot.bin

Reply:
So if you have any better ideas for the booting, just let me know, after the initial setup mess now it takes just 20 seconds to compile a new uboot :)

Reply:
Hi there,

Could you provide a containerized build environment for your U-Boot which uses an OpenWrt build toolchain? For example, here:

https://github.com/chunkeey/u-boot-ap105/pull/2

I think a full U-Boot replacement would be most reasonable if anyone can compile from source.

Reply:
(I meant an OpenWrt compiler toolchain, really*. In the linked PR, I use OpenWrt's mips compiler toolchain.)

Reply:
Nope, using 32-bit Ubuntu 14 in Virtualbox vm, with the original TP-Link sources including the toolchain.
The original FW boots perfectly with the patched uboot. So the patch method is just fine, no need of custom uboot etc.